### ISSUE TITLE
Cannot get nil value from BackingStore with delta query when a property value was changed to nil (empty)

### ISSUE THREAD

**Issue by t2y** [microsoftgraph/msgraph-sdk-go#725]
I am trying to implement the delta query to track user changes.

For example, I changed `First name` and `Last name` to empty on https://entra.microsoft.com/ .

![スクリーンショット 2024-06-06 14 21 58](https://github.com/microsoftgraph/msgraph-sdk-go/assets/96569/736fd726-8453-4701-89eb-19c560d67217)

I got the below response. You can see `givenName` and `surname` are detected as null values.

```json
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users",
  "@odata.deltaLink": "https://graph.microsoft.com/v1.0/users/delta()?$deltatoken=dqOf4D0FeifmTE8LZ08sKG_0pMPP7gZ...",
  "value": [
    {
      "givenName": null,
      "surname": null,
      "id": "a8d67d70-f83b-453d-9666-1f280a609bf6"
    }
  ]
}
```

But, I cannot know whether the `givenName` and `surname` were changed to null since the `Enumerate()` or `EnumerateKeysForValuesChangedToNil()` methods don't return the values (nil).

```go
var user models.Userable
...
b := user.GetBackingStore()
fmt.Println(b.GetInitializationCompleted())
fmt.Println(b.GetReturnOnlyChangedValues())
for i, v := range b.EnumerateKeysForValuesChangedToNil() {
	fmt.Printf(" - %v, %+v\n", i, v)
}
for k, v := range b.Enumerate() {
	fmt.Printf(" - %s, %+v\n", k, v)
}
```

```
- additionalData, map[]
- odataType, 0x140005803b0
- id, 0x14000580320
```

Another sample code to confirm the user data.

```go
json, _ := serialization.SerializeToJson(user)
fmt.Println("user", string(json))
```

Also, the user data doesn't have both `givenName` and `surname` properties.

```json
{"id":"a8d67d70-f83b-453d-9666-1f280a609bf6","@odata.type":"#microsoft.graph.user"}
```

In my debugging, `InMemoryBackingStore` doesn't handle a nil value in `User.GetFieldDeserializers()` method.

```go
res["surname"] = func(n i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.ParseNode) error {
	val, err := n.GetStringValue()
	if err != nil {
		return err
	}
	if val != nil {
		m.SetSurname(val)
	}
	return nil
}
```

```go
func (m *User) SetSurname(value *string) {
	err := m.GetBackingStore().Set("surname", value)
	if err != nil {
		panic(err)
	}
}
```

Could you tell me how do I check both `givenName` and `surname` properties were changed to nil via msgraph-sdk-go API?

## Reference

* https://learn.microsoft.com/en-us/graph/delta-query-overview
* https://github.com/microsoftgraph/msgraph-sdk-go/issues/643

**Comment by rkodev** [comment]
Hi @baywet @andrueastman This is a case where the json payload has explicitly defined a field as null, and the intention is to differentiate fields that were not returned but are with a nil value, vs fields that were explicitly set as nil because the value is null.

**Comment by baywet** [comment]
Thanks for the call out. 
This is a design limitation. The enumeration function was designed with serialization and not deserialization in mind. 
I guess we could mimic that design for deserialization with an additional enumerate read null keys method or similar.
This would be a breaking change due to the addition of a method in an interface.

Andrew, Ronald, don't hesitate to share suggestions if you have other ideas here.

**Comment by andrueastman** [comment]
One thing we could try is play around with the serialization hooks and provide and possibly provide an alternative serialization factory implementation (this would require some investigation and thinking through). 

https://github.com/microsoft/kiota-abstractions-go/blob/7677e0c3ce5056593c3c3ea6f3c7411e3e1aa42c/store/backing_store_parse_node_factory.go#L14

By having the alternative implementation have the actions to initialize the store before the json payload is deserialized(rather than after) we could have the store initialized with the values from the API in such a scenario to include nulls and other values....

**Comment by rkodev** [comment]
A pointer to an uninitialized value and a nil would evaluate to the same value but different types

```go

type Null interface{}

func checkType(v interface{}) {
	switch v := v.(type) {
	case nil:
		fmt.Println("nil value")
	case *Null:
		fmt.Println("Pointer to Null")
	default:
		fmt.Printf("Unknown type: %T\n", v)
	}
}

func main() {
	var a *Null
	var b *int

	checkType(nil) // nil value
	checkType(a)   // Pointer to Null
	checkType(b)   // Unknown type: *int

	if a == nil {
		fmt.Println("a is nil")
	}
	if a == (*Null)(nil) {
		fmt.Println("a is null")
	}
}

```
a possible solution is to define a 'Null' reference and modify kiota, abstractions and json serialization. As null value serialization should give the same result when compared to nil, but should allow handling of values explicitly set to null

**Comment by baywet** [comment]
Wouldn't any object value (non-nil) satisfy this null interface since it's empty?

**Comment by rkodev** [comment]
Yes it will, but it will give the ability to differentiate is a serialized value was null as opposed to if a serialized value was not present at all. As Is the problem we have is that there is no difference between values that were not returned by the server as opposed to values that were returned with an explicitly null value

**Comment by baywet** [comment]
Couldn't we use this instead? https://github.com/microsoft/kiota-abstractions-go/blob/main/serialization/untyped_null.go
